mesonbuild/cmake/traceparser.py

```diff
@@ -354,16 +357,16 @@ def _cmake_add_executable(self, tline: CMakeTraceLine) -> None:
         args = list(tline.args) # Make a working copy
 
         # Make sure the exe is imported
-        is_imported = True
-        if 'IMPORTED' not in args:
-            return self._gen_exception('add_executable', 'non imported executables are not supported', tline)
-
-        args.remove('IMPORTED')
-
-        if len(args) < 1:
-            return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
-
-        self.targets[args[0]] = CMakeTarget(args[0], 'EXECUTABLE', {}, tline=tline, imported=is_imported)
+        if 'IMPORTED' in args:
+            args.remove('IMPORTED')
+            if len(args) < 1:
+                return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
+            self.targets[args[0]] = CMakeTarget(args[0], None, 'EXECUTABLE', {}, tline=tline, imported=True)
+        else:
+            if len(args) < 1:
+                return self._gen_exception('add_executable', 'requires at least 1 argument', tline)
+            build_path = self.build_dir.parent / args[0]
```

Comment:
This looks a bit too fragile to me (and is probably also the cause of the current test failures). Ideally, we could use [`Executable.get_filename()`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/build.py#L1296) here, but constructing a new `Executable` looks too expensive for something that won't be needed in 90% of all cases.

The best solution I can come up with right now is to defer the construction of the `build_path` by passing a lambda instead to `CMakeTarget` and have it lazily evaluate (and cache) the result if required.